crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4640,10 +4650,413 @@ mod tests {
+                        x = 7
+                    else:
+                        x = 8
+                except OSError:
+                    x = 9
+                else:
+                    x = 10
+            else:
+                try:
+                    try:
+                        x = 11
+                    except RuntimeError:
+                        x = 12
+                    else:
+                        x = 13
+                except OSError:
+                    x = 14
+                else:
+                    x = 15
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(
+            &db,
+            "src/a.py",
+            "x",
+            "Literal[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]",
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn except_handler_multiple_definitions_in_exhaustive_nested_try_block() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                try:
+                    x = 1
+                except:
+                    x = 2
+                y = 11
+            except TypeError:
+                x = 101
+                y = 111
+            except:
+                y = 211
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(&db, "src/a.py", "x", "Unbound | Literal[1, 2, 101]");
+        assert_public_ty(&db, "src/a.py", "y", "Literal[11, 111, 211]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn multiple_definitions_of_same_symbol_in_try_block() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x = 1
+                x = 2
+                x = 3
+            except:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+        assert_public_ty(&db, "src/a.py", "x", "Unbound | Literal[1, 2, 3]");
+
+        Ok(())
+    }
+
+    #[test]
+    fn exception_raised_in_nested_try_block_caught_in_outer() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                try:
+                    x = 1
+                    x = 2
+                except TypeError:
+                    x = 3
+                else:
+                    x = 4
+            except:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // We might have left the inner `try` or the inner `except TypeError`
+        // block due to another exception, in which case the assignments in the inner `try` block
+        // would not necessarily have been overridden by the otherwise-exhaustive
+        // `except`/`else` branches in the inner block.
```

Comment:
"otherwise-exhausive except/else branches in the inner block" doesn't seem right -- those aren't exhaustive